src/extractor/guidance/intersection_handler.cpp

```diff
@@ -383,107 +383,109 @@ std::size_t IntersectionHandler::findObviousTurn(const EdgeID via_edge,
-    }
-    const std::pair<std::int64_t, std::int64_t> num_continue_names =
-        std::make_pair(continue_count, continue_count_valid);
-
-    auto sameOrHigherPriority = [&in_way](const auto &way_data) {
+    auto sameOrHigherPriority = [&in_way_data](const auto &way_data) {
         return way_data.road_classification.GetPriority() <=
-               in_way.road_classification.GetPriority();
+               in_way_data.road_classification.GetPriority();
     };
     auto IsLowPriority = [](const auto &way_data) {
         return way_data.road_classification.IsLowPriorityRoadClass();
     };
+    // These two Compare functions are used for sifting out best option and continue
+    // candidates by evaluating all the ways in an intersection by what they share
+    // with the in way. Ideal candidates are of similar road class with the in way
+    // and are require relatively straight turns.
     const auto RoadCompare = [&](const ConnectedRoad &lhs, const ConnectedRoad &rhs) {
         const EdgeData &lhs_data = node_based_graph.GetEdgeData(lhs.eid);
         const EdgeData &rhs_data = node_based_graph.GetEdgeData(rhs.eid);
         const auto lhs_deviation = angularDeviation(lhs.angle, STRAIGHT_ANGLE);
         const auto rhs_deviation = angularDeviation(rhs.angle, STRAIGHT_ANGLE);
 
-        const bool rhs_same_classification = rhs_data.road_classification == in_way.road_classification;
-        const bool lhs_same_classification = lhs_data.road_classification == in_way.road_classification;
+        const bool rhs_same_classification =
+            rhs_data.road_classification == in_way_data.road_classification;
+        const bool lhs_same_classification =
+            lhs_data.road_classification == in_way_data.road_classification;
         const bool rhs_same_or_higher_priority = sameOrHigherPriority(rhs_data);
         const bool rhs_low_priority = IsLowPriority(rhs_data);
         const bool lhs_same_or_higher_priority = sameOrHigherPriority(lhs_data);
         const bool lhs_low_priority = IsLowPriority(lhs_data);
-        auto left_tie = std::tie(
-            lhs.entry_allowed, lhs_same_or_higher_priority, rhs_low_priority, rhs_deviation, lhs_same_classification);
-        auto right_tie = std::tie(
-            rhs.entry_allowed, rhs_same_or_higher_priority, lhs_low_priority, lhs_deviation, rhs_same_classification);
+        auto left_tie = std::tie(lhs.entry_allowed,
+                                 lhs_same_or_higher_priority,
+                                 rhs_low_priority,
+                                 rhs_deviation,
+                                 lhs_same_classification);
+        auto right_tie = std::tie(rhs.entry_allowed,
+                                  rhs_same_or_higher_priority,
+                                  lhs_low_priority,
+                                  lhs_deviation,
+                                  rhs_same_classification);
         return left_tie > right_tie;
```

Comment:
what is really confusing me here is that you use `min_element` and still compare `>` here, this makes it really hard to compare what you want here.

Right now you get:

 - entry allowed before not entry allowed
 - same or higher priority before different priorities
 - all non-low priority roads before low priority roads
 - smaller deviation
 - all same deviations preferred from non same.

In the PR before you changed this, you did:

- entry allowed (due to the creation of out ways)
- deviation
- same classification (note that the find_if for same_or_higher (https://github.com/Project-OSRM/osrm-backend/pull/3217/commits/0cee3486e5de11ca2c2f191e6233392171f1317f#diff-b10c978a328e386ebdc0aa4db955388eL455) will always succeed on same_classification (contains priority)
- same or higher priority

if that did not succeed (since no same or higher priority), you selected the first one without low priority:
- entry allowed (due to the creation of out ways)
- deviation
- not low priority

While this still does not handle obviousness on service roads correctly (needs additional tests), the check worked for all other settings. Your new sorting criterion is different from your old one, as you can see from this analysis.